### ISSUE TITLE
Meeting 2019-10-22

### ISSUE THREAD

**Issue by einatnielsen** [nielsen-oss/docs#4]
## Agenda 
* https://reactjs.org/docs/testing-recipes.html - make a research about Act pattern in context to React and RTL. Obviously, go through the testing recipes.

* Snapshots - Do we want to encourage using snapshots? #6

* POC - FnComponent, ClassComponent and Redux Connected Component with RTL.
  - Functional Component - [Gist](https://gist.github.com/MatanBorenkraoutNielsen/bf867c8cbb475e2bdc40a8e05fd771a2)
  - Class Component - [Gist](https://gist.github.com/einatnielsen/bc6863ec367fef4f2a5fed763ee2b5fb)
  - Redux Connected Component - [Gist](https://gist.github.com/lidoravitan/eaf1a9f020299326e9d9b9acde2f8960)

**Comment by MatanBorenkraoutNielsen** [comment]
**Act as a pattern:**
Much needed, IMO we should follow the AAA - Arrange Act Assert pattern, with each block separated from the others to get the best clarification without diving into the test.

**Act in React:**
Actually, RTL's `fireEvent` is already wrapped in act:
```
function fireEvent(...args) {
  let returnValue
  act(() => {
    returnValue = dtlFireEvent(...args)
  })
  return returnValue
}
```
And that's the main way to interact with the DOM (changing it), I believe that we won't need to add `act` calls in our code except for specific scenarios where we will try to trigger a DOM change without `fireEvent`..

**Comment by lidoravitan** [comment]
Redux Connected Component

First I decided to go with a behavioral testing pattern, which means interacting with the rendered connected component instead of testing mapStateToPros or mapDispatchToProps.
the reason is that testing mapStateToProps isolated won't break in case we change the actual component interface(props names).

With RTL it was kind of native, I just rendered the connected container and interact with the actual dom. However, when it comes to asynchronous actions(i.e redux-logic) it fails.

With Enzyme it was difficult to make it work and not stable(versions issues) while it should be the same pattern.

Currently, I can't point that RTL is the big winner because of the async problems.
